iroh-net/src/discovery.rs

```diff
@@ -0,0 +1,93 @@
+//! Trait and utils for the node discovery mechanism.
+
+use anyhow::Result;
+use futures::stream::BoxStream;
+
+use crate::{AddrInfo, MagicEndpoint, NodeId};
+
+/// Node discovery for [`super::MagicEndpoint`].
+///
+/// The purpose of this trait is to hoop up a node discovery mechanism that
+/// allows finding information such as the derp url and current addresses
+/// of a node given the id.
+///
+/// To allow for discovery, the [`super::MagicEndpoint`] will call `publish` whenever
+/// discovery information changes. If a discovery mechanism requires a periodic
+/// refresh, it should start it's own task.
+pub trait Discovery: std::fmt::Debug + Send + Sync {
+    /// Publish the given [`AddrInfo`] to the discovery mechanisms.
+    ///
+    /// This is fire and forget, since the magicsock can not wait for successful
+    /// publishing. If publishing is async, the implementation should start it's
+    /// own task.
+    ///
+    /// This will be called from a tokio task, so it is safe to spawn new tasks.
+    /// These tasks will be run on the runtime of the [`super::MagicEndpoint`].
+    fn publish(&self, _info: &AddrInfo) {}
```

Comment:
why are these not `async`?

Reply:
I can answer this at least for the original discovery trait.

They are intended to be "fire and forget". At the site where you call these you do not want to wait for them to complete. E.g. if we publish to a HTTP PUT endpoint we don't want to wait for successful publish before continuing, since that would probably create havoc inside the magicsock and in any case not help anybody.

E.g. in case of the pkarr discovery, this just puts the record somewhere and starts a loop to publish and republish to the DHT. Publishing to the DHT can take seconds, you don't want to wait for that. And what are you going to do if it fails? E.g. you are offline.

Reply:
well from a task management perspective I would have expected a handling more like this

```rust
// in the magicsock
let publish_task = tokio::task::spawn(discovery.publish());
// track task or drop depending on the management
```

because this way you can eg cancel things on shutdown

Reply:
The task is owned by the discovery service, so it's the job of the discovery service to cancel the task on drop. I think there is not necessarily is a 1:1 relation between tasks that need to be spawned and publish calls.

E.g. in the pkarr discovery there is just a single long lived republish task owned by the discovery, and calling publish just replaces the value to be published with the new one.

Reply:
how would I communicate that I want the publish to end then? there is no shutdown method or anything like that if this is service is long running

Reply:
Should we maybe pass `CancellationToken`s into both `resolve` and `publish`? This would give full flexibility to both. Downside is that implementors would have to use `tokio::select!` likely. Still, even for the resolve case it could make implementations simpler. Depending on how a resolver works, detecting that the returned stream was dropped may not be straightforward or instant. E.g. with channels, you might only realize once you try to sent, and so you might have been doing uneeded work already up to that point.

Reply:
Or we could add an optional `fn shutdown()` to the discovery trait which will be called in `MagicEndpoint::close` and can be used by the discovery service to abort all tasks it started.

Reply:
I like the cancellation token approach

Reply:
When I saw the doc comment here I was wondering why the signature does not give you a handle to the runtime to spawn instead of having it just in the documentation that "tokio spawn is fine".

Given the cancellation discussion, maybe it could even be a `JoinSet` that you're politely asked to spawn into, and then you get aborting for free when the magicsock drops the joinset.

Reply:
Leaving this for a followup and created an issue here: https://github.com/n0-computer/iroh/issues/2066